### DISCUSSION TITLE
What about setting "Eclipse-BundlePolicy: dependent" in OSGi Metadata of jackson-databind?

### DISCUSSION THREAD

**Discussion by haubi** [FasterXML/jackson-databind#4142] [Q&A] [answered]
So I'm stumbling over some hackery here, which is about repackaging jackson-databind.jar to add "Eclipse-BundlePolicy: dependent" to OSGi Metadata. Then, that repackaged jar is put into some P2 Update Site, used to define some specific (Eclipse/OSGi) Target Platform, for the application being some Rich Client Platform application.

That metadata property is required to allow the OSGi bundle classloader of that jackson-databind bundle to find the application classes, and actually I wonder how anyone is able to use jackson-databind in some OSGi environment without that property. Somehow I don't have so much luck finding some relevant docs on that topic.

After all, I still want to add jackson-databind.jar to Target Platform, but instead of the indirection via some dedicated P2 Update Site, I want to add it as Maven dependency right within Eclipse, which is possible since ~2021.
However, above metadata hackery isn't supported there, so I would like to utilize jackson-databind in some "right way" if possible, as long as that way isn't too inconvenient.

WDYT?

Thanks!

**Comment by cowtowncoder** [comment]
Since I do not know a lot about OSGi (nor run on OSGi in production), I hope someone else can comment here. I am happy to merge PRs for improvements of course.

**Comment by haubi** [comment]
FWIW, 've found issue #624 and PR #854 about the same topic, but I haven't found a way to set up class loading via 'TypeFactory' during set up of 'ObjectMapper'.

**Reply by cowtowncoder** [reply, marked as answer]
Configuration would occur by accessing existing `TypeFactory`, calling its `withClassLoader()` method, resetting.

So either with 2.x methods of `getTypeFactory()` and `setTypeFactory()`.

Or if using `MapperBuilder` (`JsonMapper.builder()`) similar methods.

**Comment by haubi** [comment]
Meanwhile, I've learned that using `Eclipse-BundlePolicy: dependent` actually is kind of a hack against the concepts of OSGi dependency tree, potentially asking for various classloading conflicts.
Hence, there's no point for the upstream jackson-databind package to add this OSGi Metadata.
Instead, developers of OSGi applications should use `TypeFactory` and it's `withClassLoader()` method as mentioned by @cowtowncoder, thanks a lot!

**Reply by cowtowncoder** [reply]
Thank you for the investigation @haubi !
